codechecker_lib/session_manager.py

```diff
@@ -0,0 +1,299 @@
+                                toPrint = toPrint + e.message['info']
+                            if 'info' in e.message and 'desc' in e.message:
+                                toPrint = toPrint + "; "
+                            if 'desc' in e.message:
+                                toPrint = toPrint + e.message['desc']
+                        else:
+                            toPrint = e.__repr__()
+
+                        LOG.info("LDAP authentication error against " +
+                                 server["connection_url"] + " with DN: " +
+                                 query.replace("$USN$", username) + "\n" +
+                                 "Error was: " + toPrint)
+                    finally:
+                        l.unbind()
+
+            return False
+
+    def create_or_get_session(self, client, auth_string):
+        '''Create a new session for the given client and auth-string, if
+        it is valid. If an existing session is found, return that instead.'''
+        if not self.__auth_config["enabled"]:
+            return None
+
+        self.__logins_since_prune += 1
+        if self.__logins_since_prune >= \
+           self.__auth_config["logins_until_cleanup"]:
+            self.__cleanup_sessions()
+
+        if self.__handle_validation(auth_string):
+            session_already = next(
+                (s for s
+                 in sessionManager.__valid_sessions if s.client == client
+                 and s.still_reusable()
+                 and s.persistent_hash ==
+                 _Session.calc_persistency_hash(client, auth_string)),
+                None)
+
+            if session_already:
+                session_already.revalidate()
+                session = session_already
+            else:
+                # TODO: Use a more secure way for token generation?
+                token = uuid.UUID(bytes=os.urandom(16)).__str__().replace("-",
+                                                                          "")
+                session = _Session(client, token,
+                                   _Session.calc_persistency_hash(client,
+                                                                  auth_string))
+                sessionManager.__valid_sessions.append(session)
+
+            return session.token
+        else:
+            return None
+
+    def is_valid(self, client, token, access=False):
+        '''Validates a given token (cookie) against
+        the known list of privileged sessions'''
+        if not self.isEnabled():
+            return True
+        else:
+            return any(_sess.client == client
+                       and _sess.token == token
+                       and _sess.still_valid(access)
+                       for _sess in sessionManager.__valid_sessions)
+
+    def invalidate(self, client, token):
+        '''Remove a user's previous session from the store'''
+        for session in sessionManager.__valid_sessions[:]:
```

Comment:
Similar question why do we have to copy the whole __valid_sessions list if one of them is invalidated ore the sessions are cleaned up? Do we lose some session tokens if a cleanup is in progress and a new is added or checked for validity?